gobblin-data-management/src/main/java/org/apache/gobblin/data/management/source/LoopingDatasetFinderSource.java

```diff
@@ -169,6 +204,22 @@ public DeepIterator(Iterator<Dataset> baseIterator, String previousDatasetUrnWat
 
     @Override
     protected WorkUnit computeNext() {
+      if (retryWorkUnits.hasNext()) {
```

Comment:
Won't partitions that appear in the retryWorkUnits be double-processed? I think they'll be processed once when retryWorkUnits is iterated over and then processed a second time when the baseIterator is iterated over. In order to avoid double processing, I think you would need to filter out the partitions that have exceeded the max cross-execution retries limit from the dataset stream returned by the dataset finder.